strapi/strapi
# RBAC Action Aliases - v5

:warning:  **This PR targets v5. This will also be ported to v4.**

### What does it do?

#### Action Aliases

Introduce a new `aliases` property on admin actions (`RBAC`).

This property is defined as an optional array of action alias.

```ts
interface ActionAlias {
  /**
   * The action ID to alias
   */
  actionId: string;

  /**
   * An optional array of subject to restrict the alias usage
   */
  subjects?: string[];
}

type Action = {
  // other properties ...

  /**
   * An optional array of @see {@link ActionAlias}.
   *
   * It represents the possible aliases for the current action.
   *
   * Aliases are unidirectionals.
   *
   * Note: This is an internal property and probably shouldn't be used outside of Strapi core features.
   *       Its behavior might change at any time without notice.
   *
   * @private
   */
  aliases?: ActionAlias[];
};
```

#### Action Provider

A new method `aliases` is added to the admin action provider (`RBAC`).

```ts
declare function aliases(actionId: string, subject?: string | null): string[];
```

It returns an array of `actionId` that can be used as aliases for the given `actionId` and `subject`.

- When an alias doesn't declare a `subjects` property, it only checks actions ID
- When an alias has an array of subjects declared, it ensures both the actions ID and subjects matches   

#### New Aliases

A default set of aliases for common use cases have been defined:

- `admin::users.read` is an alias of `plugin::content-manager.explorer.read`(`admin::user`)
- `admin::users.roles` is an alias of `plugin::content-manager.explorer.read`(`admin::role`)
- `plugin::i18n.locale.read` is an alias of `plugin::content-manager.explorer.read`(`plugin::i18n.locale`)

### Why is it needed?

The initial idea came when looking at #19001. The issue being hidden content-types (`admin::user`, `plugin::i18n.locale`, etc...) not having a content manager action being registered for them (only CTs being displayed in the CM have one).

Since it is possible to create a relation to `admin::user` (manual exception), this is why even super admin can't see the main field for the relation (because the needed permission doesn't even exist).

**Note:** It's also already possible to declare manual relations to any internal content-type from the schema file (e.g. `locale`)

This is a limitation in how permissions are defined. However, simpler solutions like only adding the hidden content-types (mostly internal ones like `upload::file` or `upload::folder`) to the list of allowed subjects for the `plugin::content-manager.explorer.read` action might lead to unwanted side effects (like adding them to the roles' edit permission table).

The proposed (internal) `aliases` feature mitigates the issue by adding more flexible options when checking permissions while limiting side effects on already existing features and behaviors. 

### How to test it?

1. Create a content-type with a relation to `admin::user` (`CT`)
2. Create a non super-admin user (`A`), go the content manager
3. From the super-admin account (`B`), set the following permissions for (`A`)'s role
  - In `Collection Types`: allow every CRUD operation on `CT`
  - In `Settings`: make sure `Read` is **not** checked in `Users & Roles`
4. As (`A`) Go to `Content-Manager` > `CT` > `Create new entry`
5. When opening the select input for the `admin::user` relation, only document IDs should be visible
6. As (`B`), update (`A`)'s role permissions, and check the `Read` box in `Settings` > `Users & Roles
7. As (`A`), try opening the select input again, the proper main field should be displayed (by default `firstname`)

### Related issue(s)/PR(s)

#19001

### Notes

- Aliases are unidirectional to allow setting more granular permissions
- The aliases defined in this PR are just meant to address common scenarios, but this is meant to be extended based on future use cases
- While this is set as an internal feature (for now), it could be interesting to imagine the potential use cases for plugin & user app if made public

### Screenshots

![Screenshot from 2024-07-25 00-43-21](https://github.com/user-attachments/assets/97f56cb2-a9fc-4fb6-bad0-d49025c0027b)
![Screenshot from 2024-07-25 00-44-03](https://github.com/user-attachments/assets/85f294fb-5072-4762-b03b-8a755147aaab)
